File: vdf.py
#!/usr/bin/env python

# a simple parser for Valve's KeyValue format
# https://developer.valvesoftware.com/wiki/KeyValues
#
# author: Rossen Popov, 2014
#
# use at your own risk

import re

###############################################
#
# Takes a file or str and returns dict
#
# Function assumes valid VDF as input.
# Invalid VDF will result in unexpected output
#
###############################################

def parse(a):
    a_type = type(a)

    if a_type is file:
        lines = a.readlines()
    elif a_type is str:
        lines = a.split('\n')
    else:
        raise ValueError("Expected parametar to be file or str")

    obj = dict()
    stack = [obj]
    skipnext = False
    name = ""

    re_keyvalue = re.compile(r'^"((\\.|[^\"])*)"[ \t]+"((\\.|[^\"])*)(")?')

    itr = iter(lines)

    for line in itr:
        line = line.strip()

        # skip empty and comment lines
        if line == "" or line[0] == '/':
            continue

        # one level deeper
        if line[0] == "{" and skipnext is True:
            skipnext = False
            continue

        # one level back
        if line[0] == "}":
            stack.pop()
            continue

        # parse keyvalue pairs
        if line[0] == '"':
            while True:
                m = re_keyvalue.match(line)

                # we've matched a simple keyvalue pair, map it to the last dict obj in the stack
                if m:
                    # if the value is line consume one more line and try to match again, until we get the KeyValue pair
                    if m.group(5) == None:
                        line += itr.next()
                        continue

                    stack[-1][m.group(1)] = m.group(3)

                # we have a key with value in parenthesis, so we make a new dict obj (one level deep)
                else:
                    key = line[1:-1]

                    stack[-1][key] = dict()
                    stack.append(stack[-1][key])
                    skipnext = True

                # exit the loop
                break

    return obj

###############################################
#
# Take a dict, reuturns VDF in str buffer
#
# dump(dict(), pretty=True) for indented VDF
#
###############################################

def dump(a, **kwargs):
    pretty = kwargs.get("pretty", False)

    if type(pretty) is not bool:
        raise ValueError("Pretty option is a boolean")

    return _dump(a,pretty)

def _dump(a,pretty=False,level=0):
    if type(a) is not dict:
        raise ValueError("Expected parametar to be dict")

    indent = "\t"
    buf = ""
    line_indent = ""

    if pretty:
        line_indent = indent * level

    for key in a:
        if type(a[key]) is dict:
            buf += '%s"%s"\n%s{\n%s%s}\n' % (line_indent, key, line_indent, _dump(a[key],pretty,level+1), line_indent)
        else:
            buf += '%s"%s" "%s"\n' % (line_indent, key, str(a[key]))

    return buf

###############################################
#
# Testing initiative
#
###############################################


def test():
    tests = [
                [ '' , {} ],
                [ {} ,  '' ],
                [ "//comment text\n//comment", {} ],
                [ {1:1}, '"1" "1"\n'],
                [ {"a":"1","b":"2"} , '"a" "1"\n"b" "2"\n' ],
                [ '//comment\n"a" "1"\n"b" "2" //comment' , {"a":"1","b":"2"} ],
                [ {"a":{"b":{"c":{"d":"1","e":"2"}}}} , '"a"\n{\n"b"\n{\n"c"\n{\n"e" "2"\n"d" "1"\n}\n}\n}\n' ],
                [ '"a"\n{\n"b"\n{\n"c"\n{\n"e" "2"\n"d" "1"\n}\n}\n}\n' , {"a":{"b":{"c":{"d":"1","e":"2"}}}} ],
            ]

    for test,expected in tests:
        out = None

        try:
            if type(test) is dict:
                out = dump(test)
            else:
                out = parse(test)
        except:
            print "Test falure (exception):\n\n%s" % str(test)
            raise

        if cmp(expected,out) != 0:
            print "Test falure (ouput mismatch):\n\n%s" % str(test)
            print "\nOutput:\n\n%s" % str(out)
            print "\nExpected:\n\n%s\n" % str(expected)

            raise Exception("Output differs from expected result")

    return True

File: vdf.js
// a simple parser for Valve's KeyValue format
// https://developer.valvesoftware.com/wiki/KeyValues
//
// author: Rossen Popov, 2014
//
// use at your own risk

var VDF = {
    parse: function(text) {
        if(typeof text != "string") {
            throw new TypeError("VDF.parse: Expecting parametar to be a string");
        }

        lines = text.split("\n");

        var obj = {};
        var stack = [obj];
        var expect_bracket = false;
        var name = "";

        var re_keyvalue = new RegExp('^"((?:\\\\.|[^"\\\\])*)"[ \t]+"((?:\\\\.|[^"\\\\])*)(")?','m');
        var re_key = new RegExp('^"((?:\\\\.|[^\\\\"])*)"')

        var i = 0, j = lines.length;
        for(; i < j; i++) {
            line = lines[i].trim();

            // skip empty and comment lines
            if( line == "" || line[0] == '/') { continue; }

            // one level deeper
            if( line[0] == "{" ) {
                expect_bracket = false;
                continue;
            }

            if(expect_bracket) {
                throw new SyntaxError("VDF.parse: invalid syntax");
            }

            // one level back
            if( line[0] == "}" ) {
                stack.pop();
                continue;
            }

            // parse keyvalue pairs
            if( line[0] == '"' ) {
                // nessesary for multiline values
                while(true) {
                    m = re_keyvalue.exec(line);

                    // we've matched a simple keyvalue pair, map it to the last dict obj in the stack
                    if(m) {
                        // if don't match a closing quote for value, we consome one more line, until we find it
                        if(typeof m[3] == "undefined") {
                            line += "\n" + lines[++i];
                            continue;
                        }

                        stack[stack.length-1][m[1]] = m[2];
                    }
                    // we have a key with value in parenthesis, so we make a new dict obj (one level deep)
                    else {
                        m = re_key.exec(line);

                        if(!m) throw new SyntaxError("VDF.parse: invalid syntax");

                        key = m[1]

                        stack[stack.length-1][key] = {};
                        stack.push(stack[stack.length-1][key]);
                        expect_bracket = true;
                   }
                   break;
               }
            }
        }

        if(stack.length != 1) throw new SyntaxError("VDF.parse: open parentheses somewhere");

        return obj;
    },

    stringify: function(obj,pretty) {
        if( typeof obj != "object") {
                throw new TypeError("VDF.stringify: First input parameter is not an object");
        }

        pretty = ( typeof pretty == "boolean" && pretty) ? true : false;

        return this._dump(obj,pretty,0);
    },

    _dump: function(obj,pretty,level) {
        if( typeof obj != "object" ) {
            throw new TypeError("VDF.stringify: a key has value of type other than string or object");
        }

        var indent = "\t";
        var buf = "";
        var line_indent = "";


        if(pretty) {
            for(var i = 0; i < level; i++ ) {
                line_indent += indent;
            }
        }

        for(key in obj) {
            if( typeof obj[key] == "object" ) {
                buf += [line_indent, '"', key, '"\n', line_indent, '{\n', this._dump(obj[key],pretty,level+1), line_indent, "}\n"].join('');
            }
            else {
                buf += [line_indent, '"', key, '" "', String(obj[key]), '"\n'].join('');
            }
        }

        return buf;
    }
}
